### PR TITLE
Fix contrastive search to correctly handle input with padding

### PR DESCRIPTION
**Pull request by ducviet00**
# What does this PR do?

This PR fixes contrastive search to correctly handle input padding in decoder-only & encoder-decoder models.

## Details

I encountered the issue and discovered that the Contrastive Search implementation is highly sensitive to padded tokens.

For example, with the prompt: `The whispered legends of the haunted mansion spoke`, the output from Hugging Face without padding is: 
- `The whispered legends of the haunted mansion spoke of the "souls of the dead" who were "falling out of the sky" and "falling into the sea."\n`

However, with padding tokens, the output becomes: 
- `The whispered legends of the haunted mansion spoke of the "soul of the dead" and the "blood of the dead."\nThe ghost of Dr. H. P. Lovecraft was a man`

You can check the Colab notebook that demonstrates the issue [here](https://colab.research.google.com/drive/1q14xn3RpCDYxur1C1ORc_bJ60Y7m4e_n?usp=sharing)

## How did I fix the issue
The issue arises when `input_ids` contain padded tokens; the `hidden_states` of the model then include embeddings for these padded tokens if the model doesn't eliminate them during processing. The `_ranking_fast` function also calculates values based on these padded tokens, leading to incorrect outputs. This is critical because it significantly degrades the model's performance.

To fix this, I created a `cosine_matrix_mask` based on the `attention_mask` and penalized the `cosine_matrix` using this mask (ignoring padding positions) by applying large negative values. After that, I padded the `cosine_matrix_mask` with ones to match the output length.

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#create-a-pull-request),
      Pull Request section?
- [ ] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [x] Did you write any new necessary tests?

## Who can review?
@gante @ArthurZucker @amyeroberts @Rocketknight1

### CODE REVIEW

File: src/transformers/generation/utils.py [hunk 1]
```diff
@@ -2597,6 +2597,12 @@ def _contrastive_search(
         unfinished_sequences = torch.ones(batch_size, dtype=torch.long, device=input_ids.device)
         model_kwargs = self._get_initial_cache_position(input_ids, model_kwargs)
 
+        cosine_matrix_mask = None
+        if not self.config.is_encoder_decoder:
+            # Create cosine_matrix_mask based on the attention_mask
+            cosine_matrix_mask = model_kwargs["attention_mask"]
+            cosine_matrix_mask = cosine_matrix_mask.repeat_interleave(top_k, dim=0)
```
Comment (gante) [inline src/transformers/generation/utils.py:2604]: Suggestion:

```suggestion
        attention_mask_name = "decoder_attention_mask" if self.config.is_encoder_decoder else "attention_mask"
        cosine_matrix_mask = model_kwargs[attention_mask_name]
        cosine_matrix_mask = cosine_matrix_mask.repeat_interleave(top_k, dim=0)
```

That way, your correction should also work in the case where the model is an encoder-decoder model and a partially masked decoder prompt is passed :)

(we may need extra code if `attention_mask_name` doesn't exist in `model_kwargs` or `model_kwargs[attention_mask_name]` can be `None`, let's see if the tests complain)

File: src/transformers/generation/utils.py [hunk 2]
```diff
@@ -2764,7 +2770,11 @@ def _contrastive_search(
             # compute the degeneration penalty and re-rank the candidates based on the degeneration penalty and the
             # model confidence. Keeping `selected_idx` on CPU enables multi-device contrastive search and doesn't
             # introduce (noticeable) slowdowns on single-device runs.
-            selected_idx = _ranking_fast(context_hidden, next_hidden, top_k_probs, penalty_alpha, top_k)
+            selected_idx = _ranking_fast(
+                context_hidden, next_hidden, top_k_probs, cosine_matrix_mask, penalty_alpha, top_k
+            )
+            if not self.config.is_encoder_decoder:
```
Comment (gante) [inline src/transformers/generation/utils.py:2776]: (this line might need to be removed, depending on the final form of the suggestion above -- i.e. if we're able to create `cosine_matrix_mask` all the times)

File: src/transformers/generation/utils.py [hunk 3]
```diff
@@ -4287,6 +4298,14 @@ def _ranking_fast(
     norm_context_hidden = context_hidden / context_hidden.norm(dim=2, keepdim=True)
     norm_next_hidden = next_hidden / next_hidden.norm(dim=2, keepdim=True)
     cosine_matrix = torch.matmul(norm_context_hidden, norm_next_hidden.transpose(1, 2)).squeeze(-1)  # [B*K, S]
+
+    if cosine_matrix_mask is not None:
```
Comment (gante) [inline src/transformers/generation/utils.py:4302]: (this line might need to be removed, depending on the final form of the suggestion above -- i.e. if we're able to create `cosine_matrix_mask` all the times)

File: tests/generation/test_contrastive_search.py [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+
+        # Set the tokenizer to left-pad the sequences
+        tokenizer.padding_side = "left"
+
+        # Define the PAD token as the EOS token
+        tokenizer.pad_token = tokenizer.eos_token
+        model.generation_config.pad_token_id = model.generation_config.eos_token_id
+
+        # Configure the model's low_memory setting for generation
+        model.generation_config.low_memory = low_memory
```
Comment (gante) [inline tests/generation/test_contrastive_search.py:40]:
```suggestion
```

Let's not add this extra flag, to ensure we test the main code path

File: tests/generation/test_contrastive_search.py [hunk 5]
```diff
@@ -0,0 +1,84 @@
...
+            do_sample=False,
+            penalty_alpha=penalty_alpha,
+            top_k=top_k,
+            max_length=128,
+        )
+        outputs_with_padding = outputs_with_padding[:, padding_length:]
+        generated_text_with_padding = tokenizer.batch_decode(outputs_with_padding, skip_special_tokens=True)
+
+        # Assert that the generated texts are identical for padded and non-padded inputs
+        self.assertListEqual(generated_text_no_padding, generated_text_with_padding)
```
Comment (gante) [inline tests/generation/test_contrastive_search.py:84]: In addition to this check, can we also check against the expected text itself?

(that way, if contrastive search breaks in general, this test will detect it too)

File: tests/generation/test_contrastive_search.py [hunk 6]
```diff
@@ -0,0 +1,84 @@
...
+@require_torch
+class ContrastiveSearchTest(unittest.TestCase):
+    @parameterized.expand(
+        [
+            (False, 4, 0.6),
+            (True, 6, 0.4),
+        ]
+    )
+    @slow
+    def test_batch_contrastive_search_gpt2(self, low_memory, top_k, penalty_alpha):
```
Comment (gante) [inline tests/generation/test_contrastive_search.py:26]: Please move this test inside `GenerationIntegrationTests`, in `tests/generation/test_utils.py` (we don't separate tests according to the decoding strategy)

**Review by gante — COMMENTED** [review]
@ducviet00 thank you for identifying the underlying numerical issue, proposing a very reasonable fix, and implementing it with a test 💛 

I agree with the suggested fix: by masking `cosine_matrix` with a large negative value when the corresponding tokens are masked, `degeneration_penalty` will never be related to the masked tokens and, therefore, masked tokens will not have an impact on `contrastive_score`.

I've added a few minor suggestions

**Comment by ducviet00** [comment]
@gante Thanks for your feedback! I initially thought encoder-decoder models wouldn’t be affected since `decoder_input_ids` don’t seem to need padding tokens, except in continuous batching (not in this case). Applying the `cosine_matrix_mask` to these models might add some minor overhead, though. I hadn’t considered the decoder prompt before, so thanks for pointing that out! 

I’ll push an update with support for encoder-decoder models based on your suggestions soon.

File: tests/generation/test_contrastive_search.py (hunk 5 above)
Comment (ducviet00) [reply tests/generation/test_contrastive_search.py:84]: I think each model already has its own correctness test, so there's no need. 
- [test_contrastive_search_gpt2](https://github.com/huggingface/transformers/blob/605f3245dcca34381c35520c35ba0b701ed80d58/tests/models/gpt2/test_modeling_gpt2.py#L873)
- [test_contrastive_search_t5](https://github.com/huggingface/transformers/blob/1de7dc7403b3b89ec421d43a8c9ee245211a61f6/tests/models/t5/test_modeling_t5.py#L1544)
- ...

**Comment by ducviet00** [comment]
@gante 
I push an update to support encode-decoder models. I also added test for t5 model, with padding for `decoder_input_ids`. On the main branch, `outputs_with_padding`:
- `Ich muss diese Aufgabe noch vor Ende des Tages beenden.`

It should be: 
- `Ich muss diese Aufgabe vor Ende des Tages beenden.`

You can check the code inside `test_padding_input_contrastive_search_t5`, in `tests/generation/test_utils.py`

File: src/transformers/generation/utils.py [hunk 7]
```diff
@@ -2597,6 +2597,15 @@ def _contrastive_search(
...
 
+        # Create cosine_matrix_mask based on the attention_mask
+        cosine_matrix_mask = torch.ones_like(input_ids, dtype=torch.long)
+        if self.config.is_encoder_decoder:
+            if "decoder_attention_mask" in model_kwargs and model_kwargs["decoder_attention_mask"] is not None:
+                cosine_matrix_mask = model_kwargs["decoder_attention_mask"]
+        else:
+            cosine_matrix_mask = model_kwargs["attention_mask"]
+        cosine_matrix_mask = cosine_matrix_mask.repeat_interleave(top_k, dim=0)
+
```
Comment (ducviet00) [inline src/transformers/generation/utils.py:2608]: This code initializes a default mask and then updates it based on the model type.
- For encoder-decoder models, if `model_kwargs` contains a `decoder_attention_mask` (and it is not None), `cosine_matrix_mask` is set to this mask. If `decoder_attention_mask` is missing, it falls back to the default mask.
- For decoder-only models, `cosine_matrix_mask` is set to the `attention_mask` from `model_kwargs`.

Please let me know if there are any additional logic checks that need to be added

**Comment by ducviet00** [comment]
Hi @gante
PTAL 🤗

**Review by gante — APPROVED** [review]
Thank you for iterating 💛

**Review by LysandreJik — APPROVED** [review]
Very impressive change and tests! Thanks for your PR @ducviet00

**Comment by gante** [comment]
@ducviet00 thank you for making `transformers` better for everyone 💛
